<h2> Use the Kaggle Song Lyrics Dataset to identify lyrics corresponding to the indie genre for training our model.

In [0]:
import pandas as pd
import zipfile
import os
import shutil

In [2]:
from google.colab import drive
drive.mount('/content/drive/')

Go to this URL in a browser: https://accounts.google.com/o/oauth2/auth?client_id=947318989803-6bn6qk8qdgf4n4g3pfee6491hc0brc4i.apps.googleusercontent.com&redirect_uri=urn%3Aietf%3Awg%3Aoauth%3A2.0%3Aoob&scope=email%20https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fdocs.test%20https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fdrive%20https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fdrive.photos.readonly%20https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fpeopleapi.readonly&response_type=code

Enter your authorization code:
··········
Mounted at /content/drive/


In [5]:
!ls

drive  sample_data


In [0]:
#zip_data = '380000-lyrics-from-metrolyrics.zip'
unzipped = '380000-lyrics-from-metrolyrics'
#song_lyrics_path = os.path.join(os.getcwd(), unzipped)

In [0]:
#if not os.path.exists(song_lyrics_path):
#    ZIP_REF = zipfile.ZipFile(os.path.join(os.getcwd(), zip_data))
    
#    ZIP_REF.extractall()
#    ZIP_REF.close()   

In [0]:
data = pd.read_csv(os.path.join('/content/drive/My Drive/deep_learning_models/Seq_2_Seq_Models', '380000-lyrics-from-metrolyrics/lyrics.csv'))

In [10]:
data.head()

index             song  year           artist genre  \
0      0        ego-remix  2009  beyonce-knowles   Pop   
1      1     then-tell-me  2009  beyonce-knowles   Pop   
2      2          honesty  2009  beyonce-knowles   Pop   
3      3  you-are-my-rock  2009  beyonce-knowles   Pop   
4      4    black-culture  2009  beyonce-knowles   Pop   

                                              lyrics  
0  Oh baby, how you doing?\nYou know I'm gonna cu...  
1  playin' everything so easy,\nit's like you see...  
2  If you search\nFor tenderness\nIt isn't hard t...  
3  Oh oh oh I, oh oh oh I\n[Verse 1:]\nIf I wrote...  
4  Party the people, the people the party it's po...

<h4> Now look at the different genres after dropping columns with no lyrics

In [11]:
data.dropna(subset=['lyrics'])
data.describe()

index           year
count  362237.000000  362237.000000
mean   181118.000000    2008.537596
std    104568.959068       9.908024
min         0.000000      67.000000
25%     90559.000000    2006.000000
50%    181118.000000    2008.000000
75%    271677.000000    2014.000000
max    362236.000000    2038.000000

In [12]:
data.genre.value_counts()

Rock             131377
Pop               49444
Hip-Hop           33965
Not Available     29814
Metal             28408
Other             23683
Country           17286
Jazz              17147
Electronic        16205
R&B                5935
Indie              5732
Folk               3241
Name: genre, dtype: int64

<h4> There are 5732 songs which are in the indie category. Let's use these as our training data. let's see what artists are included in the indie song list

In [13]:
indie_data = data.loc[data.genre == 'Indie']
print("Unique artists = {}".format(len(set(indie_data.artist.values))))

Unique artists = 416


In [0]:
india_data = indie_data.astype({"lyrics": str})

In [0]:
lyrics_data = indie_data['lyrics'].dropna().tolist()

In [16]:
lyrics_data[1]

"Place outside where I pass to waste my time\nIn the shadows of the dark you'll find\nBehind sleepless eyes there's a woman\nTaste of my sweet\nLove like defeat can be so cruel\nIt can fill you full for a while\nYou're a lamb with fangs\nGirl dressed like crimson flames\nSo they see it\nYou're like a bad burn\nBut they never let you try\nJust to feel it\nTaste of my sweet\nLove like defeat can be so cruel\nIt can fill you full of wild dreams\nAnd the back of your knees\nCan you bend them for me please\nI could meet you for a while\nDressed so plain\nI put my head in a frame\nAnd the paints just words I scream\nWrapped up in a pretty string\nI leave a prize at the heels of your feet\nTaste of my sweet\nLove like defeat can be so cruel\nIt can fill you full of wild dreams\nAnd the back of your knees darling\nCan you bend them for me please\nI could meet you for a while"

<h3> Clean the Lyrics Data before saving

In [17]:
vocab = []
for item in range(len(lyrics_data)):
    vocab.append(sorted(set(lyrics_data[item])))
flat_vocab = sorted(set([item for sublist in vocab for item in sublist]))
print("Num characters = {} \n".format(len(flat_vocab)))
print("Unique Characters = {}".format(flat_vocab))

Num characters = 147 

Unique Characters = ['\n', ' ', '!', '"', '#', '$', '%', '&', "'", '(', ')', '*', '+', ',', '-', '.', '/', '0', '1', '2', '3', '4', '5', '6', '7', '8', '9', ':', ';', '<', '=', '>', '?', '@', 'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z', '[', ']', '^', '_', '`', 'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z', '{', '|', '}', '\x80', '\x81', '\x82', '\x83', '\x85', '\x86', '\x87', '\x88', '\x89', '\x8a', '\x8d', '\x8e', '\x91', '\x93', '\x99', '\x9a', '\x9e', '\x9f', '\xa0', '¡', '¢', '£', '¤', '¥', '¦', '§', '¨', '©', 'ª', '«', '¬', '\xad', '®', '¯', '°', '±', '²', '³', '´', '¶', '¹', 'º', '»', '¼', '½', '¾', '¿', 'Â', 'Ã', 'Å', 'â', 'ã', 'ï']


In [0]:
import numpy as np
count_chars = np.array([0*len(flat_vocab)])

for item in range(len(lyrics_data)):
    vec = []
    for char in flat_vocab:
        vec.append(lyrics_data[item].count(char))
    count_chars = count_chars +np.array(vec)


<h4> Get a sense of which characters are used and which ones are less frequently used so we can remove them from training

In [19]:
char_dict = dict(zip(flat_vocab, count_chars))
char_dict

{'\n': 92486,
 ' ': 524250,
 '!': 865,
 '"': 1148,
 '#': 4,
 '$': 1,
 '%': 1,
 '&': 33,
 "'": 34190,
 '(': 1799,
 ')': 1800,
 '*': 66,
 '+': 3,
 ',': 25580,
 '-': 1877,
 '.': 9238,
 '/': 66,
 '0': 179,
 '1': 195,
 '2': 238,
 '3': 104,
 '4': 145,
 '5': 82,
 '6': 60,
 '7': 30,
 '8': 39,
 '9': 67,
 ':': 420,
 ';': 143,
 '<': 1,
 '=': 20,
 '>': 1,
 '?': 4074,
 '@': 1,
 'A': 13628,
 'B': 5391,
 'C': 4200,
 'D': 3959,
 'E': 4692,
 'F': 2453,
 'G': 2390,
 'H': 4450,
 'I': 35383,
 'J': 1506,
 'K': 956,
 'L': 4095,
 'M': 3761,
 'N': 4438,
 'O': 6057,
 'P': 1796,
 'Q': 349,
 'R': 2481,
 'S': 8176,
 'T': 11685,
 'U': 1478,
 'V': 574,
 'W': 8755,
 'X': 66,
 'Y': 5916,
 'Z': 53,
 '[': 421,
 ']': 420,
 '^': 1,
 '_': 17,
 '`': 53,
 'a': 166535,
 'b': 31772,
 'c': 41379,
 'd': 86953,
 'e': 280886,
 'f': 38380,
 'g': 56301,
 'h': 121134,
 'i': 131837,
 'j': 4100,
 'k': 29634,
 'l': 109872,
 'm': 66257,
 'n': 164644,
 'o': 205606,
 'p': 28162,
 'q': 2231,
 'r': 118797,
 's': 127357,
 't': 192703,
 'u': 

<h4> This shows that as we expect mainly only alphanumerics are used, with characters like \x88 having only 1 use. Naturally, we should not include these characters in training. The same holds for random fractions, hashtags etc. With this, let's start cleaning the data by first running it through a regex filter to remove all the non alphabetic characters and then some.

<h5> Clean out the characters that we don't need. write a function to do this

In [88]:
import re
char_dict_as_list = list(char_dict.keys())
forbidden_chars = char_dict_as_list[:3] + [char_dict_as_list[8]]+ [char_dict_as_list[13]] + [char_dict_as_list[15]]
print(forbidden_chars)
# apply the cleaning procedure to everything except new lines and spaces.
def clean_text(text, regex):
  out_string = ''
  for item in text:
    if item in forbidden_chars:
      out_string += item
    else:
      out_string += re.sub(regex, '', item)
    
  return out_string
#return [re.sub(regex, '', item) for item in text if item not in forbidden_chars]

['\n', ' ', '!', "'", ',', '.']


In [89]:
print(lyrics_data[1][:100])
clean_text(lyrics_data[1][:100], r'[^A-Za-z]+')

Place outside where I pass to waste my time
In the shadows of the dark you'll find
Behind sleepless 


"Place outside where I pass to waste my time\nIn the shadows of the dark you'll find\nBehind sleepless "

In [0]:
cleaned_lyrics = []
for item in range(len(lyrics_data)):
    remaining_text = clean_text(lyrics_data[item], r'[^A-Za-z]+')
    remaining_text = clean_text(remaining_text, r'[^\x00-\x7F]+')
    #cleaned_lyrics.append(''.join([re.sub(r'[^\s\w_]+','', item) for item in lyrics_data[item]]))
    cleaned_lyrics.append(remaining_text)

In [91]:
vocab = []
for item in range(len(cleaned_lyrics)):
    vocab.append(sorted(set(cleaned_lyrics[item])))
flat_vocab = sorted(set([item for sublist in vocab for item in sublist]))
print("Num characters = {} \n".format(len(flat_vocab)))
print("Unique Characters = {}".format(flat_vocab))

Num characters = 58 

Unique Characters = ['\n', ' ', '!', "'", ',', '.', 'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z', 'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z']


In [92]:
cleaned_lyrics[1]

"Place outside where I pass to waste my time\nIn the shadows of the dark you'll find\nBehind sleepless eyes there's a woman\nTaste of my sweet\nLove like defeat can be so cruel\nIt can fill you full for a while\nYou're a lamb with fangs\nGirl dressed like crimson flames\nSo they see it\nYou're like a bad burn\nBut they never let you try\nJust to feel it\nTaste of my sweet\nLove like defeat can be so cruel\nIt can fill you full of wild dreams\nAnd the back of your knees\nCan you bend them for me please\nI could meet you for a while\nDressed so plain\nI put my head in a frame\nAnd the paints just words I scream\nWrapped up in a pretty string\nI leave a prize at the heels of your feet\nTaste of my sweet\nLove like defeat can be so cruel\nIt can fill you full of wild dreams\nAnd the back of your knees darling\nCan you bend them for me please\nI could meet you for a while"

<h5> A simple regex replacement reduces the number of characters in the cleaned lyrics dataset. Note that the second line in the 
loop removes control characters such as '\x...\. Now we only have alphabetic characters, lets look at a couple strings

<h5> Write the lyrics data into a text file

In [0]:
os.chdir('/content/drive/My Drive/deep_learning_models/Seq_2_Seq_Models')
with open("cleaned_lyrics.txt", 'w') as f:
    for item in cleaned_lyrics:
        f.write(repr("".join([x for x in item])))

In [100]:
# check where the file was written
print(os.getcwd())
!ls

/content/drive/My Drive/deep_learning_models/Seq_2_Seq_Models
380000-lyrics-from-metrolyrics	    main.ipynb
380000-lyrics-from-metrolyrics.zip  main.py
cleaned_lyrics.txt		    song_lyrics.txt
Extract_Song_Lyrics.ipynb	    Text_Gen_Songs.ipynb
Extract_Song_Lyrics.py		    Text_Gen_Songs.py
lyrics.txt


<h3> This generates a set of lyrics associated with indie songs which is what we would like to train our model with